Extensions/Firebase/C_firebasetools/D_storagetools.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Firebase Tools Collection
+ * @fileoverview
+ * @author arthuro555
+ */
+
+/**
+ * Firebase Storage Event Tools
+ * @namespace
+ */
+gdjs.evtTools.firebase.storage = {
+  uploads: new gdjs.UIDArray(),
+};
+
+/**
+ * Uploads a file as string to the firebase storage bucket.
+ * @param {string} file - The entire file as string.
+ * @param {string} onlinePath - The path under wich the file will be accessible on the bucket.
+ * @param {"none"|"base64"|"base64url"|"data_url"} [type] - The type/format of the string to upload.
+ * @param {gdjs.Variable} [callbackValueVariable] - The variable where to store the result (url to the file).
+ * @param {gdjs.Variable} [callbackStateVariable] - The variable where to store if the operation was successful.
+ * @param {gdjs.Variable} [callbackUIDVariable] - The variable where to store the upload ID.
+ * @param {gdjs.Variable} [callbackProgressVariable] - The variable where to store the progress.
+ */
+gdjs.evtTools.firebase.storage.upload = function (
+  file,
+  onlinePath,
+  type,
+  callbackValueVariable,
+  callbackStateVariable,
+  callbackUIDVariable,
+  callbackProgressVariable
+) {
+  type = type === 'none' ? undefined : type;
+
+  let uploadTask;
+  try {
+    uploadTask = firebase
+      .storage()
+      .ref()
+      .child(onlinePath)
+      .putString(file, type);
+  } catch (e) {
+    if (typeof callbackStateVariable !== 'undefined')
+      callbackStateVariable.setString(e.message);
+    return;
+  }
+
+  let uploadID;
+  if (typeof callbackUIDVariable !== 'undefined') {
+    // Only bother pushing if the ID will be stored.
+    uploadID = gdjs.evtTools.firebase.storage.uploads.push(uploadTask);
+    callbackUIDVariable.setNumber(uploadID);
+  }
+
+  uploadTask.on(
+    firebase.storage.TaskEvent.STATE_CHANGED,
+    (uploadProgress) => {
+      if (typeof callbackProgressVariable !== 'undefined')
+        gdjs.evtTools.network._objectToVariable(
+          uploadProgress,
+          callbackProgressVariable,
+          ['tasks'] // Remove circular reference.
```

Comment:
@arthuro555 This variable won't be used (a good reason to add typescript to these files btw, to catch such errors). I'm going to merge anyway but can you check this? :) If we need circular reference protection, we can build it in _objectToVariable.

Reply:
Indeed 🤔 I think I should do a full review of the code again as as this so a wrapper and wrapping functions is repetitive I may have overlooked more issues like this one.

Reply:
Please make a full pass again on the code - ideally adding `// @ts-check` would eliminate this category of problem entirely :)